### PR TITLE
Created digital_ocean_floating_ip module

### PR DESCRIPTION
**Pull request by pmarques**
##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

cloud/digital_ocean/digital_ocean_floating_ip

##### SUMMARY

Add support for create delete and assign Digital Ocean Floating IPs to a droplet.

### CODE REVIEW

**Comment by pmarques** [comment]
Migrated from https://github.com/ansible/ansible-modules-core/pull/5221

File: lib/ansible/modules/cloud/digital_ocean/do_floating_ip.py
Comment (abadger) [inline lib/ansible/modules/cloud/digital_ocean/do_floating_ip.py]: Note -- this PR is failing tests because this line has an extra leading plus sign.

**Comment by pmarques** [comment]
Moved facts module to new PR  https://github.com/ansible/ansible/pull/26372

File: lib/ansible/modules/cloud/digital_ocean/do_floating_ip.py
Comment (BondAnthony) [inline lib/ansible/modules/cloud/digital_ocean/do_floating_ip.py]: Spelling.. "region"

File: lib/ansible/modules/cloud/digital_ocean/do_floating_ip.py
Comment (BondAnthony) [inline lib/ansible/modules/cloud/digital_ocean/do_floating_ip.py]: Shouldn't this be absent instead of present?

File: lib/ansible/modules/cloud/digital_ocean/do_floating_ip.py
Comment (BondAnthony) [inline lib/ansible/modules/cloud/digital_ocean/do_floating_ip.py]: None of your examples below actually list the required parameter of "oauth_token". Secondly, adding oauth_token as the token parameter would create three different "api token" parameter names within the DO modules. Do you think we should pick a standard and stick with that one as we migrate the modules to the v2 api? Currently the DO modules use "api_token" and "api_key". I believe the api_key is only used in the v1 ssh keys module. Since you have written a new v2 implementation of ssh_keys should we stick with "api_token" or "oauth_token"?

File: lib/ansible/modules/cloud/digital_ocean/do_floating_ip.py
Comment (BondAnthony) [inline lib/ansible/modules/cloud/digital_ocean/do_floating_ip.py]: Other DO modules are already looking at "DO_API_TOKEN", "DO_API_KEY". Should we keep this standard as we move forward so users don't have to setup more than one environment variable?

Please see above comments about the use of oauth_token.

**Review by BondAnthony — COMMENTED** [review]
Couple questions and some documentation changes

**Comment by BondAnthony** [comment]
needs_revision

File: lib/ansible/modules/cloud/digital_ocean/do_floating_ip.py
Comment (pmarques) [reply lib/ansible/modules/cloud/digital_ocean/do_floating_ip.py]: Good points, the first I completely agree (I was using the env var and I forgot to add that one).
About the name of the **_token*. If you are asking my opinion, I called it *ouath* because is what it is although, I'm fine with keeping the old name or at least make an alias to keep backward compatibility (if this makes sense).

File: lib/ansible/modules/cloud/digital_ocean/do_floating_ip.py
Comment (pmarques) [reply lib/ansible/modules/cloud/digital_ocean/do_floating_ip.py]: nice catch!

File: lib/ansible/modules/cloud/digital_ocean/do_floating_ip.py
Comment (BondAnthony) [reply lib/ansible/modules/cloud/digital_ocean/do_floating_ip.py]: Since the newer modules coming in will be named do_* I think now would be a good time to start pushing a standard. I second your point about making the parameter oauth_token since it provides some meaning to the end users. Since your adding a number of new modules and we only have three working DO modules today it will be easier to update the existing three to oauth_token.

Just wanting to know what you were thinking since you have a number of PRs. Thank you for contributing!

File: lib/ansible/modules/cloud/digital_ocean/do_floating_ip.py
Comment (BondAnthony) [reply lib/ansible/modules/cloud/digital_ocean/do_floating_ip.py]: Probably best to include the older DO_API_TOKEN env as an additional failback.

File: lib/ansible/modules/cloud/digital_ocean/do_floating_ip.py
Comment (pmarques) [reply lib/ansible/modules/cloud/digital_ocean/do_floating_ip.py]: Sure, I can create a new PR to add the alias to **oauth_token** to the old modules.

**Review by BondAnthony — APPROVED** [review]

**Comment by BondAnthony** [comment]
shipit

@abadger up for you to review.

File: lib/ansible/modules/cloud/digital_ocean/do_floating_ip.py
Comment (pmarques) [reply lib/ansible/modules/cloud/digital_ocean/do_floating_ip.py]: I just noticed that old module use **DO_API_KEY** and **DO_API_TOKEN**, so I added **DO_API_KEY** to keep the previous ones.

**Review by BondAnthony — APPROVED** [review]

**Comment by BondAnthony** [comment]
@pmarques I would assume you will be changing the name of this module based on the changes that occurred on #18930?

**Comment by pmarques** [comment]
@BondAnthony yes, actually I need to change other PRs too (#26365  #26372), just waiting for @ryansb to approve.

**Comment by pmarques** [comment]
btw, this modules still tagged with **affects_2.3** should it be updated to **affects_2.4**

**Comment by pmarques** [comment]
@BondAnthony change all my PRs to match **digital_ocean_***

**Comment by BondAnthony** [comment]
-label affects_2.3

**Comment by BondAnthony** [comment]
+label affects_2.4

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_floating_ip.py
Comment (BondAnthony) [inline lib/ansible/modules/cloud/digital_ocean/digital_ocean_floating_ip.py]: Mispelling of "region"

**Comment by BondAnthony** [comment]
shipit

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_floating_ip.py
Comment (abadger) [inline lib/ansible/modules/cloud/digital_ocean/digital_ocean_floating_ip.py]: catchall exception handling isn't very useful as it hides whatever might have gone wrong from sight.  I'd probably just call core(module) here.   If there is a specific class of errors that you are expecting, then catching those and returning a more informative error message could be appropriate.  I often see the following pattern used which isn't so bad:
``` python
import traceback

from ansible.module_utils._text import to_native

try:
    core(module)
except SomeException as e:
    module.fail_json(msg="Helpful error message: %s" % to_native(e), exception=traceback.format_exc())
```

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_floating_ip.py
Comment (abadger) [inline lib/ansible/modules/cloud/digital_ocean/digital_ocean_floating_ip.py]: I see that the requirements lists python-2.6 as the minimum python version.  Unfortunately, format strings in python-2.6 are more limited than in python-2.7 and later.  So you need to be explicit about the position in the format args list you are looking at like this:
``` python
response = rest.get('floating_ips/{0}/actions/{1}'.format(ip, action_id))
```

There's a few other places with format strings that have to be fixed as well.

**Comment by abadger** [comment]
Added a few comments of things to fix.  Once those are done this can be merged.  Feel free to ping me to merge this via IRC if you want.  I'm abadger1999 in #ansible-devel on irc.freenode.net.

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_floating_ip.py
Comment (pmarques) [reply lib/ansible/modules/cloud/digital_ocean/digital_ocean_floating_ip.py]: Ups, my bad. I forgot to change this in some modules. Should be fixed.

File: lib/ansible/modules/cloud/digital_ocean/digital_ocean_floating_ip.py
Comment (pmarques) [reply lib/ansible/modules/cloud/digital_ocean/digital_ocean_floating_ip.py]: True, I should be using the same behaviour on PR #18930, just calling core(module)

**Comment by abadger** [comment]
merged to devel for 2.4.0  Thanks @pmarques and @BondAnthony!

**Pull request merged**
